### PR TITLE
Add new model, Tranformer, implemented by Pytorch's ``TransformerEncoder``.

### PR DESCRIPTION
**Pull request by mingjii**
Transformer for language model.

### CODE REVIEW

File: lmp/model/__init__.py [hunk 1]
```diff
@@ -48,6 +48,7 @@
 from lmp.model._sattn_gru import SAttnGRUBlock, SAttnGRUModel
 from lmp.model._sattn_lstm import SAttnLSTMBlock, SAttnLSTMModel
 from lmp.model._sattn_rnn import SAttnRNNBlock, SAttnRNNModel
+from lmp.model._transformer import TransformerModel, PositionalEncoding
```
Comment (allenoxli) [inline lmp/model/__init__.py]: run `isort lmp` to lint `import` statements

File: lmp/model/_transformer.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
+r"""Transformer Language Model with Transformer's encoder architecture."""
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Use `nn.TransformerDecoderLayer` instead

File: lmp/model/_transformer.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+        """
+        # Add positional encoding to each sequences.
+        # Input shape : `(B, S, H)`
+        # Output shape : `(B, S, H)`
+        pe = self.pe.detach().to(src.device)
+        output = src + pe[:, :src.size(1)]
+        return self.dropout(output)
+
+
+class TransformerModel(RNNModel):
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Since `Transformer` is not an `RNN` model, use `BaseModel` as base class instead.

File: lmp/model/_transformer.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+            n_hid_lyr=n_hid_lyr,
+            n_post_hid_lyr=n_post_hid_lyr,
+            n_pre_hid_lyr=n_pre_hid_lyr,
+            p_emb=p_emb,
+            p_hid=p_hid,
+            tknzr=tknzr,
+            **kwargs,
+        )
+
+        if d_hid % n_head != 0:
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Enforce type check first for parameters (including `n_head`, `max_seq_len`, etc).

File: lmp/model/_transformer.py [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+
+class PositionalEncoding(nn.Module):
+    r"""Positional Encoding.
+
+    In order for the model to make use of the order of the sequence,
+    add ``positional encodings`` to the input embeddings.
+
+    .. math::
+
+            \begin{align}
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Only need one indentation for each `.rst` block.

File: lmp/model/_transformer.py [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+import torch.nn as nn
+
+from lmp.model._rnn import RNNModel
+from lmp.tknzr._base import BaseTknzr
+
+
+class PositionalEncoding(nn.Module):
+    r"""Positional Encoding.
+
+    In order for the model to make use of the order of the sequence,
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Change to
```
    In order for Transformer model to make use of postional information of
    input sequence, we usually add positional encodings to the input
    embeddings.
```

File: lmp/model/_transformer.py [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+
+    Parameters
+    ==========
+    d_hid: int
+        Hidden dimension for MLP and Transformer.
+        Must be bigger than or equal to ``1``.
+    dropout: float
+        Dropout probability.
+        Must satisfy ``0.0 <= dropout <= 1.0``.
+    max_len: int
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Enforce variables name consistency (change to `max_seq_len`).

File: lmp/model/_transformer.py [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+
+    Attributes
+    ==========
+    dropout: torch.nn.Dropout
+        Positional Encoding dropout.
+    pe: torch.nn.FloatTensor
+        The values of Positional Encoding.
+    """
+
+    def __init__(self, d_hid, dropout, max_len):
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Define function with keywords only parameters and add type annotations.
(change to `def __init__(self, *, d_hid: int, dropout: float, max_seq_len: int):`)

File: lmp/model/_transformer.py [hunk 9]
```diff
@@ -0,0 +1,460 @@
...
+        Must satisfy ``0.0 <= dropout <= 1.0``.
+    max_len: int
+        Max length of the input sequence.
+        Must be bigger than or equal to ``1``.
+
+    Attributes
+    ==========
+    dropout: torch.nn.Dropout
+        Positional Encoding dropout.
+    pe: torch.nn.FloatTensor
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: `torch.FloatTensor` is an utility function which return `torch.Tensor` with `dtype == torch.float64`.
(change to `torch.Tensor`)

File: lmp/model/_transformer.py [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+
+    def __init__(self, d_hid, dropout, max_len):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        # Create positional encoding table.
+        # Shape : `(S, H)`
+        self.pe = torch.zeros(max_len, d_hid)
+
+        # Position order from `0` to `S - 1`.
```
Comment (allenoxli) [inline lmp/model/_transformer.py:73]: Add shape information.

File: lmp/model/_transformer.py (hunk 5 above)
Comment (allenoxli) [reply lmp/model/_transformer.py]: You use `d_model` in the formulation but use `d_hid` throughout your code instead.
Be consistent.

File: lmp/model/_transformer.py [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+        self.pe[:, 1::2] = torch.cos(position * div_term)
+        self.pe = self.pe.unsqueeze(0)
+
+    def forward(self, src: torch.Tensor) -> torch.Tensor:
+        r"""Perform forward pass.
+
+        Parameters
+        ==========
+        src: torch.Tensor
+            Input sequence with shape ``(B, S, E)``.
```
Comment (allenoxli) [inline lmp/model/_transformer.py:91]: Use `H` or `E` but not both to denote hidden dimension.

File: lmp/model/_transformer.py [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+        # Input tensor : Output of `self.pre_hid`.
+        # Input shape  : `(B, S, H)`.
+        # Input dtype  : `torch.float32`.
+        # Output tensor: Batch of sequences with Positional Encoding.
+        # Output shape : `(B, S, H)`.
+        # Output dtype : `torch.float32`.
+        self.pe = PositionalEncoding(d_hid, p_hid, max_seq_len)
+
+        # A sigle layer architecture of Transformer encoder.
+        self.encoderlayer = nn.TransformerEncoderLayer(d_hid, n_head)
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Missing `d_ff` (feed forward layer dimension) and dropout.

File: lmp/model/_transformer.py [hunk 13]
```diff
@@ -0,0 +1,460 @@
...
+        Llion Jones, Aidan N. Gomez, Lukasz Kaiser, Illia Polosukhin.
+        2017. Attention is all you need.
+    """
+    model_name: ClassVar[str] = 'Transformer'
+
+    def __init__(
+            self,
+            *,
+            n_head: int,
+            d_emb: int,
```
Comment (allenoxli) [inline lmp/model/_transformer.py:297]: Redundant parameters.

File: lmp/model/_transformer.py [hunk 14]
```diff
@@ -0,0 +1,460 @@
...
+        self.encoderlayer = nn.TransformerEncoderLayer(d_hid, n_head)
+
+        # Stack of `n_hid_lyr` encoder layers.
+        # Input tensor : Output of `self.pe`.
+        # Input shape  : `(S, B, H)`.
+        # Input dtype  : `torch.float32`.
+        # Output tensor: Batch of Transformer encoder output.
+        # Output shape : `(S, B, H)`.
+        # Output dtype : `torch.float32`.
+        self.tranformerencoder = nn.TransformerEncoder(
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Missing `nn.LayerNorm` instance.

File: lmp/model/_transformer.py [hunk 15]
```diff
@@ -0,0 +1,460 @@
...
+        Forward pass algorithm is structured as follow:
+
+        #. Input batch of previous token ids.
+           (shape: ``(B, S)``)
+        #. Use batch of previous token ids to perform token embeddings lookup
+           on ``self.emb``.
+           (shape: ``(B, S, E)``)
+        #. Use ``self.emb_dp`` to drop some features in token embeddings.
+           (shape: ``(B, S, E)``)
+        #. Use ``self.pre_hid`` to transform token embeddings from embedding
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Wrong comments.

File: lmp/model/_transformer.py [hunk 16]
```diff
@@ -0,0 +1,460 @@
...
+
+        # Token embedding dropout.
+        # Input  shape: `(B, S, E)`.
+        # Output shape: `(B, S, E)`.
+        batch = self.emb_dp(batch)
+
+        # Transform from embedding dimension to hidden dimension.
+        # Input  shape: `(B, S, E)`.
+        # Output shape: `(B, S, H)`.
+        batch = self.pre_hid(batch)
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Transformer do not have embedding to hidden dimension projection layers.

File: lmp/model/_transformer.py [hunk 17]
```diff
@@ -0,0 +1,460 @@
...
+
+        # Transform to the origin shape.
+        # Input  shape: `(S, B, H)`.
+        # Output shape: `(B, S, H)`.
+        batch = batch.transpose(0, 1).contiguous()
+
+        # Transform from hidden dimension to embedding dimension.
+        # Input  shape: `(B, S, H)`.
+        # Output shape: `(B, S, E)`.
+        batch = self.post_hid(batch)
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Transformer does not have hidden to embedding projection layer.

File: lmp/model/_transformer.py [hunk 18]
```diff
@@ -0,0 +1,684 @@
...
+        \text{PE}_{(\text{pos},\text{2i})} &= \sin\
+            (\frac{\text{pos}}{10000^\frac{\text{2i}}{d_{\text{emb}}}}) \\
+        \text{PE}_{(\text{pos},\text{2i+1})} &= \cos\
+            (\frac{\text{pos}}{10000^\frac{\text{2i}}{d_{\text{emb}}}}) \\
+        \end{align}
+
+    Parameters
+    ==========
+    d_emb: int
+        embden dimension for MLP and Transformer.
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: The first letter of a sentence should be a capital letter. And it should be `Embedding` instead.

File: lmp/model/_transformer.py [hunk 19]
```diff
@@ -0,0 +1,684 @@
...
+        2017. Attention is all you need.
+    """
+
+    def __init__(
+        self,
+        *,
+        d_emb: int,
+        n_head: int,
+        p_hid: float,
+        d_ff: int,
```
Comment (allenoxli) [inline lmp/model/_transformer.py:151]: Sort parameter by alphabet order.

File: lmp/model/_transformer.py [hunk 20]
```diff
@@ -0,0 +1,684 @@
...
+        n_hid_lyr: int,
+        **kwargs: Optional[Dict],
+    ):
+        super().__init__()
+
+        # A sigle layer architecture of Transformer encoder,
+        # TransformerEncoderLayer is made up of a ``MultiHeadAttention`` layer
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: Do not need to use ``.

File: lmp/model/_transformer.py [hunk 21]
```diff
@@ -0,0 +1,684 @@
...
+        2017. Attention is all you need.
+    """
+    model_name: ClassVar[str] = 'Transformer'
+
+    def __init__(
+            self,
+            *,
+            n_head: int,
+            d_emb: int,
+            d_ff: int,
```
Comment (allenoxli) [inline lmp/model/_transformer.py:298]: Sort parameter by alphabet order.

File: lmp/model/_transformer.py [hunk 22]
```diff
@@ -0,0 +1,684 @@
...
+            n_hid_lyr: int,
+            p_emb: float,
+            p_hid: float,
+            max_seq_len: int,
+            tknzr: BaseTknzr,
+            **kwargs: Optional[Dict],
+    ):
+        super().__init__()
+
+        if not isinstance(n_head, int):
```
Comment (allenoxli) [inline lmp/model/_transformer.py:313]: Instance check for other parameter like `d_emb`.

File: lmp/model/_transformer.py [hunk 23]
```diff
@@ -0,0 +1,684 @@
...
+        # Target dtype : `torch.int64`.
+        # Output tensor: Average next tokens prediction loss.
+        # Output shape : `(1)`.
+        # Output dtype : `torch.float32`.
+        return F.cross_entropy(logits, batch_next_tkids)
+
+    def pred(self, batch_prev_tkids: torch.Tensor) -> torch.Tensor:
+        r"""Next token prediction.
+
+        Use forward pass ouput logits to choose the most possible token id
```
Comment (allenoxli) [inline lmp/model/_transformer.py]: It should be `output` instead.

**Review by yiting-tom — COMMENTED** [review]
LGTM!
